python/GafferUI/Backups.py

```
@@ -0,0 +1,305 @@
+#
+##########################################################################
+
+import Gaffer
+
+import IECore
+
+from Qt import QtCore
+
+import collections
+import os
+import stat
+import weakref
+
+class Backups( object ) :
+
+	def __init__( self, applicationRoot ) :
+
+		# \todo I wonder if it would make sense for this sort
+		# of "bolt on" component to be a GraphComponent that was
+		# parented under the ApplicationRoot somewhere, instead
+		# of manually tracking the "parent" like this?
+		self.__applicationRoot = weakref.ref( applicationRoot )
+
+		self.__settings = Gaffer.Plug()
+		self.__settings["enabled"] = Gaffer.BoolPlug( defaultValue = True )
+		self.__settings["frequency"] = Gaffer.IntPlug( defaultValue = 5, minValue = 0 )
+		self.__settings["fileName"] = Gaffer.StringPlug( defaultValue = "${script:directory}/.gafferBackups/${script:name}-backup${backup:number}.gfr" )
+		self.__settings["files"] = Gaffer.IntPlug( defaultValue = 10, minValue = 1 )
+
+		applicationRoot["preferences"]["backups"] = self.__settings
+
+		self.__plugSetConnection = applicationRoot["preferences"].plugSetSignal().connect( self.__plugSet )
+
+		self.__timer = QtCore.QTimer()
+		self.__timer.timeout.connect( Gaffer.WeakMethod( self.__timeout ) )
+
+		# Initialise the timer
+		self.__plugSet( self.__settings["enabled"] )
+
+	# Creates a backup for the specified script, returning the
+	# name of the backup file. This is really only public so we
+	# can call it from the unit tests.
+	def backup( self, script ) :
+
+		fileName = None
+		existingFiles = []
+		for f in self.__potentialFileNames( script ) :
+			if not os.path.exists( f ) :
+				fileName = f
+				break
+			else :
+				existingFiles.append( ( os.path.getmtime( f ), f ) )
+
+		if fileName is None :
+			# All files exist already, sort by modification
+			# time and choose the oldest.
+			existingFiles.sort()
+			fileName = existingFiles[0][1]
+
+		dirName = os.path.dirname( fileName )
+		try :
+			os.makedirs( dirName )
```

I think shutil has a solution that doesn't raise when the directorty exists

---

That would be great - do you know what it is exactly? I couldn't see anything in https://docs.python.org/2/library/shutil.html.

---

Gah, I was wrong. It was copytree and rmtree that have the extra flag for ignoring errors. Looks like python3.2+ might help, but I guess we're hosed until then.